MCS-1301: Enable Unity Top-Down Plotter, and Set Screenshot Folder, via Python Config

alexcraneunity opened the pull request:
This is the Unity component of the [MCS-1301 task](https://nextcentury.atlassian.net/browse/MCS-1301). 
Python changes will have a separate pull request.

alexcraneunity commented on unity/Assets/Scripts/TopDownPlotter.cs:
```diff
@@ -53,15 +39,30 @@ public void TakeScreenshot()
 
     }
 
+    // The following methods/coroutines are left here in case we want to take screenshots on certain frame intervals in the future.
```
I left these in here for future extensibility/different use cases for screenshot taking, but I can remove them as a part of this PR as well.

kdrumm commented on unity/Assets/Scripts/AgentManager.cs:
```diff
@@ -1482,6 +1482,8 @@ public class ServerAction
 	public int straight = 1;
 	public Transform agentTransform;
 	public bool restrictOpenDoors = false;
+	public string screenshotsLocation;
+	public bool enableScreenshots;
```
I think it would be good to rename 'screenshot' to something more description 'topDownScreenshot' or 'topDownImage'.  When I think screenshot, I would think the normal visual image that the user/AI sees.

alexcraneunity replied:
Fixed to "topDownImage".

kdrumm commented on unity/Assets/Scripts/MCSController.cs:
```diff
@@ -76,6 +76,9 @@ private enum HapticFeedback {
     public Dictionary<string, SimObjPhysics> agentObjectAssociations = new Dictionary<string, SimObjPhysics>();
     private MCSMain mcsMain;
 
+    private bool recordScenes;
```
same rename here.  maybe 'recordTopDown'

alexcraneunity replied:
Fixed to "recordTopDown".

ThomasSchellenbergNextCentury commented:
@alexcraneunity Is this PR missing commits? I see the python PR is setting properties like `recordTopDown` and `topDownImagePath`, but this PR is still showing the old names.

alexcraneunity commented:
@ThomasSchellenbergNextCentury Good catch, thanks! I missed pushing the latest commit. You can now see the correct variable names.

ThomasSchellenbergNextCentury reviewed (commented):
Looking great to me! Two minor comments, up for discussion:

1. The top-down images get saved in the `Screenshots` subfolder (e.g. if I run a scene named "thomas_scene_1", the images are in "thomas_scene_1/Screenshots"). It's not super important, but should we rename this folder to be more descriptive of its specific content -- maybe something like "TopDownScreenshots"?
2. The first image (with the "_0" suffix) is always the ceiling of the default 10x10 room. Is there any way it could show the actual scene, after initialization but before the first action/step, and without the ceiling(*)? If not, I think we should just delete it entirely (or never save it in the first place), since it's a bit misleading currently.

(*) I'm not sure exactly why this is happening, since you're taking a screenshot at the end of ProcessControlCommand, which should include the Initialize action. Maybe you need to yield a frame between processing the Initialize action and taking the screenshot? Not sure.

kdrumm commented:
I agree with thomas's comments, but otherwise everything works well for me.

kdrumm reviewed (commented):
Removed.

kdrumm commented:
I think the unity portion should just write the files in the folder provided from python.  Python can add the `TopDownScreenshots` folder if desired (I like it).  I also think we should also put `topdown` in the filename so it is something like `playroom_topdown_1.png`.  @ThomasSchellenbergNextCentury , what do you think?

brianpippin commented:
> I think the unity portion should just write the files in the folder provided from python. Python can add the `TopDownScreenshots` folder if desired (I like it). I also think we should also put `topdown` in the filename so it is something like `playroom_topdown_1.png`. @ThomasSchellenbergNextCentury , what do you think?

I fully support adding "topdown" to the image names so it is clear.

kdrumm commented:
> @kdrumm That sounds reasonable, and would eliminate the need to specifically clear the "Screenshots" folder, as you mentioned in the MCS PR.

I was intending to keep a subfolder, but scrapping it would just be simpler.

alexcraneunity commented:
Thank you @kdrumm @ThomasSchellenbergNextCentury and @brianpippin. I have changed the screenshot directory, added a default one for the case when you're testing in the editor, and fixed the first image bug (by waiting a frame as @ThomasSchellenbergNextCentury suggested).
